pubtools/tracing.py

```diff
@@ -0,0 +1,123 @@
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import \
+    OTLPSpanExporter
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+log = logging.getLogger(__name__)
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if not os.getenv("PUB_OTEL_TRACING", "").lower() == "true":
+            return None
+
+        if TracingWrapper.__instance is None:
+            log.info("Creating TracingWrapper instance")
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
+                endpoint=f"{os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT')}/v1/traces",
+            )
+            cls.provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: os.getenv("OTEL_SERVICE_NAME")})
+            )
+            cls.processor = BatchSpanProcessor(otlp_exporter)
+            cls.provider.add_span_processor(cls.processor)
+            trace.set_tracer_provider(cls.provider)
+            set_global_textmap(propagator)
+            cls.tracer = trace.get_tracer(__name__)
+        return cls.__instance
+
+
+def instrument_func(span_name=None, args_to_attr=False):
+    """Instrument tracing for a function.
+
+    Args:
+        span_name: str
+            Span name.
+        args_to_attr: boolean
+            Add function parameters into span attributes.
+
+    Returns:
+        The decorated function or class
+    """
+    tracer = trace.get_tracer(__name__)
+
+    def _instrument_func(func):
+        @functools.wraps(func)
+        def wrap(*args, **kwargs):
+            if not os.getenv("PUB_OTEL_TRACING", "").lower() == "true":
+                return func(*args, **kwargs)
+
+            trace_ctx = None
+            token = None
+            # Try to extract trace context from environment variables.
+            if not context.get_current():
+                trace_ctx = propagator.extract(carrier=os.environ)
+                trace_ctx = baggage_propagator.extract(
+                    carrier=os.environ, context=trace_ctx
+                )
+
+            if trace_ctx:
+                token = context.attach(trace_ctx)
+
+            attributes = {
+                "function_name": func.__qualname__,
+            }
+            if args_to_attr:
```

Comment:
I would rather this is moved up to the top before the `PUB_OTEL_TRACING` check, even though the result is not used when it's false.

The reason is that there's no guarantee the arguments can be converted to string successfully.  I think we can say it's a bug to use  args_to_attr=True in that case, but the problem is that the bug will not be detected unless  `PUB_OTEL_TRACING` is set.  So, this sets up a trap where somebody adds `@instrument_func` on some new code and it seems to work fine when they test locally with `PUB_OTEL_TRACING` not set, but as soon as it's deployed into production where `PUB_OTEL_TRACING` is set, it crashes.

By moving up this attribute calculation to the beginning, we ensure bugs in the attribute calculation will be detected immediately even if tracing is disabled.

Reply:
Updated as your suggestion.